### PR TITLE
Add Internal-, ExternalConnection

### PR DESCRIPTION
**Pull request by Bolodya1997**
## Description
1. Adds new types:
```go
type ExternalConnection struct {
	Connection
}

func (c *ExternalConnection) IsExternal() bool {
	return true
}

type InternalConnection struct {
	Connection
}

func (c *InternalConnection) IsExternal() bool {
	return false
}
```
2. Edits `DialContext` and `StartAndDialContext` to return `*ExternalConnection` and `*InternalConnection`.

## Motivation
When we create memif socket using abstract socket, we need to additionally pass application net NS to VPP. We have the following cases:
1. By location:
    1. VPP is located in the same net NS with application (internal VPP case).
    2. VPP is located in some other net NS (external VPP case).
2. By privileges:
    1. VPP has privileged access.
    2. VPP has no privileged access.

We can pass to VPP either `""` or `"file"` as a net NS:
1. `[1, 1]` - works correct.
3. `[1, 2, ""]` - works correct.
4. `[1, 2, "file"]` - fails with net NS change error.
5. `[2, 1, ""]` - creates socket in VPP net NS, incorrect.
6. `[2, 1, "file"]` - works correct.
7. `[2, 2]` - invalid setup, doesn't make sense.

In short:
1. `[1, *, ""]` - always work.
2. `[2, *, "file"]` - always work (if setup is valid).

So we need a way to distinguish internal VPP case from external VPP case.

### CODE REVIEW

**Comment by Bolodya1997** [comment]
@edwarnicke 
cc

**Comment by Bolodya1997** [comment]
My fault, have totally missed `golangci` checks :)
Currently fixed.

**Comment by edwarnicke** [comment]
Or we could just check to see if /proc/self/ns/net is the same inode as "file" ...

**Comment by Bolodya1997** [comment]
> Or we could just check to see if /proc/self/ns/net is the same inode as "file" ...

What should we do in the case when `/proc/self/ns/net` is equal to `file`?
1. In **external** case we still need to pass `file`, because VPP doesn't know anything about the application net NS.
2. In **internal** case we can pass `""`. We can also pass `file`, but only in cases when VPP is running in privileged mode.

So it is actually the case why do we need `isExternal` flag.

**Comment by edwarnicke** [comment]
It seems to me we have realistically three cases:

NSC
NSE
Forwarder

If the NSE and NSC are what you call 'internal' and the [NSC , NSE ] mechanisms are:
1. [^memif, memif] - then the Forwarder has the NetNS and abstract socket filename from the NSE, and should pass it to the VPP instance, whether internal or external
2. [memif,^memif] - then the Forwarder has the NetNS from the NSC, and is deciding the filename itself... and so has both to pass to the VPP instance
3. [memif,memif] - directmemif, in which case the Forwarder is not using its VPP at all

So am I correct that this is primarily a problem of a hypthetical NSE or NSC wanting to use an external VPP instance?  Not the Forwarder?

**Comment by Bolodya1997** [comment]
> So am I correct that this is primarily a problem of a hypthetical NSE or NSC wanting to use an external VPP instance? Not the Forwarder?

Yes. Currently there is another option that can be used - `WithChangeNetNS` flag that Forwarder sets for `memif.{Server|Client}`. We can check for this flag and:
1. If it is set - pass net NS file to the VPP.
2. If it is not set - pass "" to the VPP.

But it will lead us to the invalid behavior if NSC/NSE uses external VPP instance.

@edwarnicke 
What do you think of using `WithChangeNetNS` instead of changes from this PR?
